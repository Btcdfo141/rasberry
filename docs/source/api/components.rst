:mod:`spencerassistant.components`
===============================

air\_quality
--------------------------------------------

.. automodule:: spencerassistant.components.air_quality
    :members:
    :undoc-members:
    :show-inheritance:

alarm\_control\_panel
--------------------------------------------

.. automodule:: spencerassistant.components.alarm_control_panel
    :members:
    :undoc-members:
    :show-inheritance:

binary\_sensor
--------------------------------------------

.. automodule:: spencerassistant.components.binary_sensor
    :members:
    :undoc-members:
    :show-inheritance:

camera
---------------------------

.. automodule:: spencerassistant.components.camera
    :members:
    :undoc-members:
    :show-inheritance:

calendar
---------------------------

.. automodule:: spencerassistant.components.calendar
    :members:
    :undoc-members:
    :show-inheritance:

climate
---------------------------

.. automodule:: spencerassistant.components.climate
    :members:
    :undoc-members:
    :show-inheritance:

conversation
---------------------------

.. automodule:: spencerassistant.components.conversation
    :members:
    :undoc-members:
    :show-inheritance:

cover
---------------------------

.. automodule:: spencerassistant.components.cover
    :members:
    :undoc-members:
    :show-inheritance:

device\_tracker
---------------------------

.. automodule:: spencerassistant.components.device_tracker
    :members:
    :undoc-members:
    :show-inheritance:

fan
---------------------------

.. automodule:: spencerassistant.components.fan
    :members:
    :undoc-members:
    :show-inheritance:

light
---------------------------

.. automodule:: spencerassistant.components.light
    :members:
    :undoc-members:
    :show-inheritance:

lock
---------------------------

.. automodule:: spencerassistant.components.lock
    :members:
    :undoc-members:
    :show-inheritance:

media\_player
---------------------------

.. automodule:: spencerassistant.components.media_player
    :members:
    :undoc-members:
    :show-inheritance:

notify
---------------------------

.. automodule:: spencerassistant.components.notify
    :members:
    :undoc-members:
    :show-inheritance:

remote
---------------------------

.. automodule:: spencerassistant.components.remote
    :members:
    :undoc-members:
    :show-inheritance:

switch
---------------------------

.. automodule:: spencerassistant.components.switch
    :members:
    :undoc-members:
    :show-inheritance:

sensor
-------------------------------------

.. automodule:: spencerassistant.components.sensor
    :members:
    :undoc-members:
    :show-inheritance:

vacuum
-------------------------------------

.. automodule:: spencerassistant.components.vacuum
    :members:
    :undoc-members:
    :show-inheritance:

water\_heater
-------------------------------------

.. automodule:: spencerassistant.components.water_heater
    :members:
    :undoc-members:
    :show-inheritance:

weather
---------------------------

.. automodule:: spencerassistant.components.weather
    :members:
    :undoc-members:
    :show-inheritance:

webhook
---------------------------

.. automodule:: spencerassistant.components.webhook
    :members:
    :undoc-members:
    :show-inheritance:
