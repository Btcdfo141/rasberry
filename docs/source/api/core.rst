.. _core_module:

:mod:`spencerassistant.core`
-------------------------

.. automodule:: spencerassistant.core
    :members: