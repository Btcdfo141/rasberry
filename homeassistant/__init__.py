"""Init file for Home Assistant."""


test2
