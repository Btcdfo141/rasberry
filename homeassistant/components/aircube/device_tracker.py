"""Support for Ubiquiti airCube routers."""
import json
import logging
import re

import requests
import voluptuous as vol

from homeassistant.components.device_tracker import (
    DOMAIN,
    PLATFORM_SCHEMA,
    DeviceScanner,
)
from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_USERNAME, CONF_VERIFY_SSL
import homeassistant.helpers.config_validation as cv

_LOGGER = logging.getLogger(__name__)

DEFAULT_VERIFY_SSL = True

PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
    {
        vol.Required(CONF_HOST): cv.string,
        vol.Required(CONF_PASSWORD): cv.string,
        vol.Required(CONF_USERNAME): cv.string,
        vol.Optional(CONF_VERIFY_SSL, default=DEFAULT_VERIFY_SSL): cv.boolean,
    }
)


def get_scanner(hass, config):
    """Validate the configuration and return an airCube scanner."""
    scanner = AirCubeDeviceScanner(config[DOMAIN])

    return scanner if scanner.success_init else None


def _refresh_on_access_denied(func):
    """If session timed out, need new session_id."""

    def decorator(self, *args, **kwargs):
        """Wrap the function to refresh session_id on error."""
        try:
            return func(self, *args, **kwargs)
        except (
            PermissionError,
            KeyError,
            NameError,
            requests.exceptions.ConnectionError,
        ):
            _LOGGER.debug(
                "Possible authentication error."
                "Trying to refresh session_id and rerun."
            )
            self.session_id = _get_session_id(
                self.url, self.username, self.password, self.verify_ssl
            )
            return func(self, *args, **kwargs)

    return decorator


class AirCubeDeviceScanner(DeviceScanner):
    """
    This class queries an Ubiquiti airCube wireless router.

    Script adapted from ubus (OpenWrt) device tracker.
    """

    def __init__(self, config):
        """Initialize the scanner."""
        host = config[CONF_HOST]
        self.username = config[CONF_USERNAME]
        self.password = config[CONF_PASSWORD]
        self.verify_ssl = config[CONF_VERIFY_SSL]

        self.parse_api_pattern = re.compile(r"(?P<param>\w*) = (?P<value>.*);")
        self.last_results = {}
        self.url = f"https://{host}/ubus"

        self.session_id = _get_session_id(
            self.url, self.username, self.password, self.verify_ssl
        )
        self.clients = []
        self.mac2name = None
        self.success_init = self.session_id is not None

    def scan_devices(self):
        """Scan for new devices and return a list with found device IDs."""
        self._update_info()
        return self.last_results

    def _generate_mac2name(self):
        """Return empty MAC to name dict."""
        self.mac2name = dict()

    @_refresh_on_access_denied
    def get_device_name(self, device):
        """Return the name of the given device or None if we don't know."""
        if self.mac2name is None:
            self._generate_mac2name()
        if self.mac2name is None:
            return None
        name = self.mac2name.get(device.upper(), None)
        return name

    @_refresh_on_access_denied
    def _update_info(self):
        """
        Ensure the information from the router is up to date.

        Returns boolean if scanning successful.
        """
        if not self.success_init:
            return False

        _LOGGER.info("Checking for clients")

        self.last_results = []
        results = 0

        result = _req_json_rpc(
            self.url, self.session_id, "call", "ubnt", "stats", self.verify_ssl
        )

        mac_data0 = result["result"][1]["results"]["wireless"]["interface"]["wlan0"][
            "assoclist"
        ]
        for i in mac_data0:
            self.last_results.append(i["mac"])
            results = results + 1
        # Only airCube AC has 5GHz radio
        try:
            mac_data1 = result["result"][1]["results"]["wireless"]["interface"][
                "wlan1"
            ]["assoclist"]
            for i in mac_data1:
                self.last_results.append(i["mac"])
                results = results + 1
        except KeyError:
            pass

        return bool(results)


def _req_json_rpc(url, session_id, rpcmethod, subsystem, method, verify_ssl, **params):
    """Perform one JSON RPC operation."""
    data = json.dumps(
        {
            "jsonrpc": "2.0",
            "id": 1,
            "method": rpcmethod,
            "params": [session_id, subsystem, method, params],
        }
    )

    try:
        res = requests.post(url, data=data, timeout=5, verify=verify_ssl)

    except (requests.exceptions.Timeout):
        _LOGGER.warning(
            "Connection to airCube timed out. Check that IP address is correct"
            "and that device is accessible."
        )
    except (requests.exceptions.SSLError):
        _LOGGER.warning("SSL error. Set VERIFY_SSL to False.")
    except (requests.exceptions.ConnectionError):
        _LOGGER.warning("Error connecting to airCube.")
    else:
        return res.json()


def _get_session_id(url, username, password, verify_ssl):
    """Get the authentication token for the given host+username+password."""
    res = []
    try:
        res = _req_json_rpc(
            url,
            "00000000000000000000000000000000",
            "call",
            "session",
            "login",
            verify_ssl=verify_ssl,
            username=username,
            password=password,
        )
    finally:
        pass
    if res is not None:
        try:
            return res["result"][1]["ubus_rpc_session"]
        except IndexError:
            return _LOGGER.warning("Authentication error: Check username and password.")
