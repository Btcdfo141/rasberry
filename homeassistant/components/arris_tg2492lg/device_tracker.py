"""Support for Arris TG2492LG router."""
import logging

from arris_tg2492lg import ConnectBox
import voluptuous as vol

from homeassistant.components.device_tracker import (
    DOMAIN,
    PLATFORM_SCHEMA,
    DeviceScanner,
)
from homeassistant.const import CONF_HOST, CONF_PASSWORD
import homeassistant.helpers.config_validation as cv

_LOGGER = logging.getLogger(__name__)

DEFAULT_IP = "http://192.168.178.1"

PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
    {
        vol.Required(CONF_PASSWORD): cv.string,
        vol.Optional(CONF_HOST, default=DEFAULT_IP): cv.string,
    }
)


def get_scanner(hass, config):
    """Return the Arris device scanner."""
    conf = config[DOMAIN]
    connect_box = ConnectBox(conf[CONF_HOST], conf[CONF_PASSWORD])
    return ArrisDeviceScanner(connect_box)


class ArrisDeviceScanner(DeviceScanner):
    """This class queries a Arrus TG2492LG router for connected devices."""

    def __init__(self, connect_box):
        """Initialize the scanner."""
        self.connect_box: ConnectBox = connect_box

    def scan_devices(self):
        """Scan for new devices and return a list with found device IDs."""
        devices = self.connect_box.get_connected_devices()

        def filter_online(device):
            return device.online is True

        def map_mac_address(device):
            return device.mac

        # filter online devices
        online_devices = list(filter(filter_online, devices))

        # show only mac address
        mac_addresses = list(map(map_mac_address, online_devices))

        # remove duplicates as some devices are returned with ipv4 and ipv6
        return list(set(mac_addresses))

    def get_device_name(self, device):
        """Return the name of the given device or None if we don't know."""
        return None
