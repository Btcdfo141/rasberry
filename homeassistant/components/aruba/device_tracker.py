"""Support for Aruba Access Points."""
import json
import logging
import re

import requests
from requests import Session
import pexpect
import voluptuous as vol

from homeassistant.components.device_tracker import (
    DOMAIN,
    PLATFORM_SCHEMA,
    DeviceScanner,
)
from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_USERNAME
import homeassistant.helpers.config_validation as cv

_LOGGER = logging.getLogger(__name__)

_DEVICES_REGEX = re.compile(
    r"(?P<name>([^\s]+)?)\s+"
    + r"(?P<ip>([0-9]{1,3}[\.]){3}[0-9]{1,3})\s+"
    + r"(?P<mac>([0-9a-f]{2}[:-]){5}([0-9a-f]{2}))\s+"
)

PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
    {
        vol.Required(CONF_HOST): cv.string,
        vol.Required(CONF_PASSWORD): cv.string,
        vol.Required(CONF_USERNAME): cv.string,
    }
)

DEFAULT_USE_API = False
DEFAULT_TIMEOUT = 10
USE_API = False
login_result_sid = ''

def get_scanner(hass, config):
    """Validate the configuration and return a Aruba scanner."""
    scanner = ArubaDeviceScanner(config[DOMAIN])

    return scanner if scanner.success_init else None


class ArubaDeviceScanner(DeviceScanner):
    """This class queries a Aruba Access Point for connected devices."""

    def __init__(self, config):
        """Initialize the scanner."""
        self.host = config[CONF_HOST]
        self.username = config[CONF_USERNAME]
        self.password = config[CONF_PASSWORD]

        global USE_API
        global login_result_sid

        self.last_results = {}

        # Test if we can use REST API
        # Aruba Instant API only likes double quotes in the json data and rejects single quotes
        login_data = '{"user": "' + self.username+ '", "passwd": "'+ self.password +'" }'
        login_headers = {'Content-Type': 'application/json' }
        login_method = "POST"
        login_resource = "https://" + self.host + ":4343/rest/login"

        rest_login = RestData(
            login_method,
            login_resource,
            login_headers,
            login_data,
            False,
            DEFAULT_TIMEOUT
        )
        rest_login.update()

        if rest_login.headers is not None:
            login_result_content_type = rest_login.headers.get("content-type")
            if login_result_content_type == "application/json":
                login_result_json = json.loads(rest_login.data)
                login_result_status = login_result_json.get("Status")
                if login_result_status == "Success":
                    USE_API = True
                    login_result_sid = login_result_json.get("sid")
                    _LOGGER.debug("Successfully logged in to Aruba Instant REST API")
                else:
                    USE_API = DEFAULT_USE_API
                    _LOGGER.debug("Error logging in to Aruba Instant REST API")

        data = self.get_aruba_data()
        self.success_init = data is not None

    def scan_devices(self):
        """Scan for new devices and return a list with found device IDs."""
        self._update_info()
        return [client["mac"] for client in self.last_results]

    def get_device_name(self, device):
        """Return the name of the given device or None if we don't know."""
        if not self.last_results:
            return None
        for client in self.last_results:
            if client["mac"] == device:
                return client["name"]
        return None

    def _update_info(self):
        """Ensure the information from the Aruba Access Point is up to date.

        Return boolean if scanning successful.
        """
        if not self.success_init:
            return False

        data = self.get_aruba_data()
        if not data:
            return False

        self.last_results = data.values()
        return True

    def get_aruba_data(self):
        """Retrieve data from Aruba Access Point and return parsed result."""

        if USE_API == True:
            _LOGGER.debug("Using REST API")
            show_clients_data = ''
            show_clients_headers = {'Content-Type': 'application/json' }
            show_clients_method = "GET"
            show_clients_resource = "https://" + self.host + ":4343/rest/show-cmd?iap_ip_addr=" + self.host + "&cmd=show%20clients&sid=" + login_result_sid

            rest_show_clients = RestData(
                show_clients_method,
                show_clients_resource,
                show_clients_headers,
                show_clients_data,
                False,
                DEFAULT_TIMEOUT
            )
            rest_show_clients.update()

            if rest_show_clients.headers is not None:
                show_clients_result_content_type = rest_show_clients.headers.get("content-type")
                if show_clients_result_content_type == "application/json":
                    show_clients_result_json = json.loads(rest_show_clients.data)
                    show_clients_result_status = show_clients_result_json.get("Status")
                    if show_clients_result_status == "Success":
                        show_clients_result_command_output = show_clients_result_json.get("Command output")
                        show_clients_result_bytes = (json.dumps(show_clients_result_command_output)).encode()
                        devices_result = show_clients_result_bytes.split(b"\\n")
                    else:
                        _LOGGER.debug("Error getting client list via Aruba Instant REST API")
                        return

        else:
            _LOGGER.debug("Using SSH")
            connect = f"ssh {self.username}@{self.host}"
            ssh = pexpect.spawn(connect)
            query = ssh.expect(
                [
                    "password:",
                    pexpect.TIMEOUT,
                    pexpect.EOF,
                    "continue connecting (yes/no)?",
                    "Host key verification failed.",
                    "Connection refused",
                    "Connection timed out",
                ],
                timeout=120,
            )
            if query == 1:
                _LOGGER.error("Timeout")
                return
            if query == 2:
                _LOGGER.error("Unexpected response from router")
                return
            if query == 3:
                ssh.sendline("yes")
                ssh.expect("password:")
            elif query == 4:
                _LOGGER.error("Host key changed")
                return
            elif query == 5:
                _LOGGER.error("Connection refused by server")
                return
            elif query == 6:
                _LOGGER.error("Connection timed out")
                return
            ssh.sendline(self.password)
            ssh.expect("#")
            ssh.sendline("show clients")
            ssh.expect("#")
            devices_result = ssh.before.split(b"\r\n")
            ssh.sendline("exit")

        devices = {}
        for device in devices_result:
            match = _DEVICES_REGEX.search(device.decode("utf-8"))
            if match:
                devices[match.group("ip")] = {
                    "ip": match.group("ip"),
                    "mac": match.group("mac").upper(),
                    "name": match.group("name"),
                }
        return devices

class RestData:
    """Class for handling the data retrieval."""

    def __init__(
        self, method, resource, headers, data, verify_ssl, timeout=DEFAULT_TIMEOUT
    ):
        """Initialize the data object."""
        self._method = method
        self._resource = resource
        self._headers = headers
        self._request_data = data
        self._verify_ssl = verify_ssl
        self._timeout = timeout
        self._http_session = Session()
        self.data = None
        self.headers = None

    def __del__(self):
        """Destroy the http session on destroy."""
        self._http_session.close()

    def update(self):
        """Get the latest data from REST service with provided method."""
        _LOGGER.debug("Updating from %s", self._resource)
        try:
            response = self._http_session.request(
                self._method,
                self._resource,
                headers=self._headers,
                data=self._request_data,
                timeout=self._timeout,
                verify=self._verify_ssl,
            )
            self.data = response.text
            self.headers = response.headers
        except requests.exceptions.RequestException as ex:
            _LOGGER.error("Error fetching data: %s failed with %s", self._resource, ex)
            self.data = None
            self.headers = None
