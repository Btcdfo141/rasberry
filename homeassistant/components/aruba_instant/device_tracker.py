import logging

from typing import Dict
import async_timeout
from datetime import timedelta

from homeassistant.components.device_tracker.config_entry import ScannerEntity
from homeassistant.components.device_tracker.const import SOURCE_TYPE_ROUTER
from homeassistant.helpers import device_registry as dr
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed

from .const import DOMAIN

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(hass, config_entry, async_add_entities):
    """Set up device tracker for Aruba Instant component."""
    device_registry = await dr.async_get_registry(hass)
    virtual_controller = hass.data[DOMAIN][config_entry.entry_id]
    await virtual_controller.async_setup()
    add_devices(virtual_controller, config_entry, device_registry)

    async def async_update_data():
        """Fetch data from API endpoint.

        This is the place to pre-process the data to lookup tables
        so entities can quickly look up their data.
        """
        try:
            async with async_timeout.timeout(10):
                return await virtual_controller.async_update_clients()
        except Exception as err:
            raise UpdateFailed(f"Error communicating with API: {err}")

    coordinator = DataUpdateCoordinator(
        hass,
        _LOGGER,
        name="aruba_instant_tracker",
        update_method=async_update_data,
        update_interval=timedelta(seconds=30),
    )
    await coordinator.async_refresh()
    async_add_entities(
        (
            InstantClientEntity(coordinator, idx, ent)
            for idx, ent in enumerate(coordinator.data)
        ),
        True,
    )


def add_devices(instant, config_entry, device_registry):
    """Add APs into HA as devices."""
    access_points = instant.aps
    for ap in access_points:
        device_registry.async_get_or_create(
            config_entry_id=config_entry.entry_id,
            connections={(dr.CONNECTION_NETWORK_MAC, access_points[ap]["ip"])},
            identifiers={(DOMAIN, access_points[ap]["serial"])},
            manufacturer="Aruba Networks",
            name=access_points[ap]["name"],
            model=access_points[ap]["type"],
        )


class InstantClientEntity(ScannerEntity):
    """Instant Client Entity"""
    def __init__(self, coordinator, idx, ent):
        self.coordinator = coordinator
        self.idx = idx
        self._mac = ent
        self._name = coordinator.data[self._mac].get("name")
        self._ip = coordinator.data[self._mac].get("ip")
        self._mac = coordinator.data[self._mac].get("mac")
        self._os = coordinator.data[self._mac].get("os")
        self._essid = coordinator.data[self._mac].get("essid")
        self._channel = coordinator.data[self._mac].get("channel")
        self._phy = coordinator.data[self._mac].get("phy")
        self._role = coordinator.data[self._mac].get("role")
        self._ipv6 = coordinator.data[self._mac].get("ipv6")
        self._signal = coordinator.data[self._mac].get("signal")
        self._signal_text = coordinator.data[self._mac].get("signal_text")
        self._speed = coordinator.data[self._mac].get("speed")
        self._speed_text = coordinator.data[self._mac].get("speed_text")

    @property
    def should_poll(self):
        """No need to poll. Coordinator notifies entity of updates."""
        return False

    @property
    def available(self):
        """Return if entity is available."""
        return self.coordinator.last_update_success

    async def async_added_to_hass(self):
        """When entity is added to hass."""
        self.async_on_remove(
            self.coordinator.async_add_listener(self.async_write_ha_state)
        )

    async def async_turn_on(self, **kwargs):
        """Turn the light on.

        Example method how to request data updates.
        """

        await self.coordinator.async_request_refresh()

    async def async_update(self):
        """Update the entity.

        Only used by the generic entity update service.
        """
        await self.coordinator.async_request_refresh()

    @property
    def is_connected(self):
        """Return the status of a client on the network."""
        # TODO research a way to determine status
        return True

    @property
    def source_type(self) -> str:
        """Return the source type."""
        return SOURCE_TYPE_ROUTER

    @property
    def unique_id(self) -> str:
        """Return client mac address as unique id."""
        return self._mac

    @property
    def name(self) -> str:
        """Return name of client."""
        return self._name

    @property
    def device_state_attributes(self) -> Dict[str, any]:
        """Return device info."""
        return {
            "name": self._name,
            "ip": self._ip,
            "mac": self._mac,
            "os": self._os,
            "essid": self._essid,
            "channel": self._channel,
            "phy": self._phy,
            "role": self._role,
            "ipv6": self._ipv6,
            "signal": self._signal,
            "signal_text": self._signal_text,
            "speed": self._speed,
            "speed_text": self._speed_text,
        }
