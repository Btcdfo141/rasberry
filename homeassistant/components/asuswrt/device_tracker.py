"""Support for ASUSWRT routers."""
import logging

from aioasuswrt.asuswrt import AsusWrt

from homeassistant.components.device_tracker import (
    DOMAIN,
    PLATFORM_SCHEMA,
    DeviceScanner,
)

import voluptuous as vol

from homeassistant.const import (
    CONF_HOST,
    CONF_MODE,
    CONF_PASSWORD,
    CONF_PORT,
    CONF_PROTOCOL,
    CONF_USERNAME,
)

from homeassistant.helpers import config_validation as cv

DEFAULT_SSH_PORT = 22
DEFAULT_INTERFACE = "eth0"
DEFAULT_DNSMASQ = "/var/lib/misc"

CONF_DNSMASQ = "dnsmasq"
CONF_INTERFACE = "interface"
CONF_PUB_KEY = "pub_key"
CONF_REQUIRE_IP = "require_ip"
CONF_SENSORS = "sensors"
CONF_SSH_KEY = "ssh_key"
SECRET_GROUP = "Password or SSH Key"

FIRST_RETRY_TIME = 60
MAX_RETRY_TIME = 900

_LOGGER = logging.getLogger(__name__)

PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
    {
        vol.Required(CONF_HOST): cv.string,
        vol.Required(CONF_USERNAME): cv.string,
        vol.Optional(CONF_PROTOCOL, default="ssh"): vol.In(["ssh", "telnet"]),
        vol.Optional(CONF_MODE, default="router"): vol.In(["router", "ap"]),
        vol.Optional(CONF_PORT, default=DEFAULT_SSH_PORT): cv.port,
        vol.Optional(CONF_REQUIRE_IP, default=False): cv.boolean,
        vol.Exclusive(CONF_PASSWORD, SECRET_GROUP): cv.string,
        vol.Exclusive(CONF_SSH_KEY, SECRET_GROUP): cv.isfile,
        vol.Exclusive(CONF_PUB_KEY, SECRET_GROUP): cv.isfile,
        vol.Optional(CONF_INTERFACE, default=DEFAULT_INTERFACE): cv.string,
        vol.Optional(CONF_DNSMASQ, default=DEFAULT_DNSMASQ): cv.string,
    }
)


async def async_get_scanner(hass, config):
    """Validate the configuration and return an ASUS-WRT scanner."""
    return AsusWrtDeviceScanner(config[DOMAIN])


class AsusWrtDeviceScanner(DeviceScanner):
    """This class queries a router running ASUSWRT firmware."""

    def __init__(self, config):
        """Initialize the scanner."""
        self.last_results = {}
        self.success_init = False
        self.config = config
        self.connection = AsusWrt(
            config[CONF_HOST],
            config[CONF_PORT],
            config[CONF_PROTOCOL] == "telnet",
            config[CONF_USERNAME],
            config.get(CONF_PASSWORD, ""),
            config.get("ssh_key", config.get("pub_key", "")),
            config[CONF_MODE],
            config[CONF_REQUIRE_IP],
            interface=config[CONF_INTERFACE],
            dnsmasq=config[CONF_DNSMASQ],
        )

        self._connect_error = False

    async def async_scan_devices(self):
        """Scan for new devices and return a list with found device IDs."""
        await self.async_update_info()
        return list(self.last_results.keys())

    async def async_get_device_name(self, device):
        """Return the name of the given device or None if we don't know."""
        if device not in self.last_results:
            return None
        return self.last_results[device].name

    async def async_get_extra_attributes(self, mac):
        """Return extra attributes to report in the GUI."""
        return {CONF_HOST: self.config[CONF_HOST]}

    async def async_update_info(self):
        """Ensure the information from the ASUSWRT router is up to date.

        Return boolean if scanning successful.
        """
        _LOGGER.debug("Checking Devices")

        try:
            self.last_results = await self.connection.async_get_connected_devices()
            if self._connect_error:
                self._connect_error = False
                _LOGGER.info("Reconnected to ASUS router for device update")
        except OSError as err:
            if not self._connect_error:
                self._connect_error = True
                _LOGGER.error(
                    "Error connecting to ASUS router for device update: %s", err
                )
