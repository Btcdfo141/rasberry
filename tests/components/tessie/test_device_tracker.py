"""Test the Tessie device tracker platform."""

from syrupy import SnapshotAssertion

from homeassistant.components.device_tracker import DOMAIN as DEVICE_TRACKER_DOMAIN
from homeassistant.core import HomeAssistant

from .common import setup_platform


async def test_device_tracker(hass: HomeAssistant, snapshot: SnapshotAssertion) -> None:
    """Tests that the device tracker entities are correct."""

    assert len(hass.states.async_all(DEVICE_TRACKER_DOMAIN)) == 0

    await setup_platform(hass)

    assert hass.states.async_all(DEVICE_TRACKER_DOMAIN) == snapshot(name="all")
