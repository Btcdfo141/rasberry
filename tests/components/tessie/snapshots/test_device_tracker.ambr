# serializer version: 1
# name: test_device_tracker[all]
  list([
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'friendly_name': 'Test',
        'gps_accuracy': 0,
        'heading': 185,
        'latitude': -30.222626,
        'longitude': -97.6236871,
        'source_type': <SourceType.GPS: 'gps'>,
        'speed': None,
      }),
      'context': <ANY>,
      'entity_id': 'device_tracker.test',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': 'not_home',
    }),
    StateSnapshot({
      'attributes': ReadOnlyDict({
        'friendly_name': 'Test Route',
        'gps_accuracy': 0,
        'latitude': 30.2226265,
        'longitude': -97.6236871,
        'source_type': <SourceType.GPS: 'gps'>,
      }),
      'context': <ANY>,
      'entity_id': 'device_tracker.test_route',
      'last_changed': <ANY>,
      'last_updated': <ANY>,
      'state': 'not_home',
    }),
  ])
# ---
